libs/s25main/mapGenerator/Map.h

```diff
@@ -18,133 +18,164 @@
 #ifndef Map_h__
 #define Map_h__
 
-#include "gameTypes/MapCoordinates.h"
+#include "mapGenerator/FeatureMap.h"
+#include "mapGenerator/TextureMapping.h"
+#include "mapGenerator/Triangles.h"
+#include "mapGenerator/RandomUtility.h"
+#include "gameData/WorldDescription.h"
+#include "helpers/containerUtils.h"
+#include "libsiedler2/archives.h"
+
 #include <string>
-#include <vector>
+#include <set>
 
-namespace libsiedler2 {
-class Archiv;
-}
+namespace rttr {
+namespace mapGenerator {
 
-using VecUChar = std::vector<unsigned char>;
+struct Features
+{
+    libsiedler2::Animal animal;
+    libsiedler2::Resource resource;
+    libsiedler2::ObjectInfo objectInfo;
+    libsiedler2::ObjectType objectType;
+    
+    Height z;
+    TexturePair texture;
+    
+    Features();
+    Features(Texture defaultTexture);
+};
 
-/**
- * Data type for reading, writing and generating maps.
- */
-struct Map
+class Map : public FeatureMap<Features>
```

Comment:
Hey @Flamefire if you have some time, could you have a quick look at this comment, please? I just want to avoid going the wrong direction. So currently, I try to move the index computation into a single place (by overriding MapBase class). Therefore I added a template type `FeatureMap<TFeature>` which is basically a `MapBase` which stores a single value per MapPoint (maybe ValueMap is a better name?). It will (eventually) lead to remove many classes/functions, like Brush/BrushSettings, Tile & GridUtilities - basically all which are doing weird calculation based on a wrong understanding of the geometry (https://settlers2.net was giving some useful insight what the grid is actually looking like). Does this look okay so far?

Anyway, now I had two possible approaches - either I keep references to multiple FeatureMap types (option 1) within the Map class or the Map class itself inherits from FeatureMap (as currently implemented, option 2). I personally prefer option 1) - even though 2 looked a bit nicer initially. What do you think?

Reply:
This sounds a lot like `NodeMapBase`, have you seen that? Name might not be optimal either but it does what you want and should be equal to your current `FeatureMap`. You can still have structs as your "nodes" or single values.

Your question is 2 fold:
- Store 1 value/node and have N maps for the N features or Store N values/node in 1 map
- Inheritance or composition

The first is mainly an optimization but also allows to limit the influence of code parts. If you need to work with 1 feature in isolation (e.g. only the animals, then only the resources) then it makes sense to have N maps. It will also be faster due to caching (check "AoS vs SoA"). (S2 does this: the "layers" in the map). Just check your interfaces what is best, e.g. `adjustHeights(heightMap)` or `adjustHeights(heightAnimalsResourcesAndEverythingMap)` is easy to decide what you need, but I guess there is also `setResources(terrain, height, resources)` where individual maps are a bit bulky. But having those individual maps you can set some of them `const` and others not when passing to functions, which might be easier to read and reason about. But the `MapBase` contains also `GetPointsInRadius` and `CheckPointsInRadius` which won't really work when you have multiple maps and need to check conditions on multiple ones (e.g "texture is mountain and height is bigger than 5"), but that could be improved

I wouldn't use multiple references. Use multiple instances.

About the second part: Inheriting from multiple feature maps is not possible (well... not useful), so if you go for that, the question is answered. If you go for 1 featuremap: Decide if the map "has a" featuremap or "is-a" featuremap from a logicial PoV. It really depends how you are working with it.

Side note: The `Map_` class looks pretty much like the `libsiedler2::ArchivItemMap`, maybe you can use or wrap that or the `glArchivItemMap` (which is badly named and should be merged/moved to `libsiedler2::ArchivItemMap`)

Ok, as you see I can't give a clear recommendation. Just check the options you have, the (intended) usage and then decide. You'll learn something either way and in the end both ways work.

Reply:
Haven't seen the `NodeMapBase ` class - yeah it is exactly the same. 

I probably wasn't clear about my point, but my question was just about 1 value / node or N values / node. I'll go for 1 value / node:
* seems to me like the better choice since you can function inputs can be specific (const)
* and does not always require the entire `Map` instance
* CheckPointsInRadius` is a valid point, but I'll see how I can work around it - `GetPointsInRadius` returns MapPoints anyway (assuming all maps have the same size it would still work). 

For the `Map_` class it's going to be deleted. This was just an intermediate commit and I didn't want to break everything, so I just renamed it for now to keep things compiling.

Reply:
About `CheckPointsInRadius`: My idea is to make that a free function which takes a variadic argument of `NodeMapBase`s and will call your functor with 1 value (ref) of each map. Only problem there is the default values, so they don't get mixed up with the variadic stuff